ui/components/app/snaps/snap-permissions-list/snap-permissions-list.js

```diff
@@ -1,52 +1,128 @@
-import React from 'react';
+import React, { useState, useEffect, useMemo } from 'react';
 import PropTypes from 'prop-types';
 import { useSelector } from 'react-redux';
-import { getWeightedPermissions } from '../../../../helpers/utils/permission';
 import { useI18nContext } from '../../../../hooks/useI18nContext';
-import { Box } from '../../../component-library';
+import { Box, ButtonLink } from '../../../component-library';
 import {
   getMultipleTargetsSubjectMetadata,
   getSnapsMetadata,
 } from '../../../../selectors';
+import {
+  Display,
+  FlexDirection,
+  JustifyContent,
+} from '../../../../helpers/constants/design-system';
+import SnapPermissionAdapter from '../snap-permission-adapter';
+import { PermissionWeightThreshold } from '../../../../../shared/constants/permissions';
 import { getSnapName } from '../../../../helpers/utils/util';
-import SnapPermissionCell from '../snap-permission-cell';
+import { getWeightedPermissions } from '../../../../helpers/utils/permission';
 
 export default function SnapPermissionsList({
   snapId,
   snapName,
   permissions,
   connections,
   showOptions,
+  turnOffAbstraction,
+  showAllPermissions,
 }) {
   const t = useI18nContext();
-  const snapsMetadata = useSelector(getSnapsMetadata);
-  const permissionsToShow = {
-    ...permissions,
-    connection_permission: connections ?? {},
-  };
+
+  const combinedPermissions = useMemo(() => {
+    return { ...permissions, connection_permission: connections ?? {} };
+  }, [permissions, connections]);
+
   const targetSubjectsMetadata = useSelector((state) =>
     getMultipleTargetsSubjectMetadata(state, connections),
   );
 
-  return (
-    <Box className="snap-permissions-list">
-      {getWeightedPermissions({
+  const snapsMetadata = useSelector(getSnapsMetadata);
+
+  const [showAll, setShowAll] = useState(false);
+
+  const [permissionsToDisplay, setPermissionsToDisplay] = useState([]);
```

Comment:
I mostly agree with what @hmalik88 is saying here. I don't think we should be storing the permissions in state.

IMO, we should only store the `showAll` flag in state and then conditionally choose the permissions set to display.

It's not clear to me why this would cause performance problems @david0xd 🤔

Reply:
Forgive the simplification, but what prevents us from doing something like:

```js
const [showAll, setShowAll] = useState(false);

// Optionally memoize filteredPermissions if the calculation is expensive
const permissionsToShow = showAll ? permissions : filteredPermissions;

```

Reply:
Updated implementation with new approach.